flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java

```diff
@@ -0,0 +1,323 @@
+                createPartitionSplit(
+                        topicName, 0, Boundedness.CONTINUOUS_UNBOUNDED, MessageId.latest);
+
+        reader.addSplits(Collections.singletonList(emptySplit));
+
+        TestingReaderOutput<Integer> output = new TestingReaderOutput<>();
+        InputStatus status;
+        status = reader.pollNext(output);
+        // We sleep for 1 sec to avoid entering false negative case:
+        // The fetcher could become idle upon entry because :
+        // 1. the AddSplitsTask is dequeued from taskQueue,
+        // 2. The dequeued splits is not assigned to assignedSplits.
+        // And if both assignedSplits is null and taskQueue is empty, the fetcher could become idle,
+        // and return NOTHING_AVAILABLE upon entry. This happened multiple times during local test.
+        sleepUninterruptibly(1, TimeUnit.SECONDS);
+        assertThat(status).isEqualTo(InputStatus.NOTHING_AVAILABLE);
+        reader.close();
+    }
+
+    protected void setupSourceReader(
+            PulsarSourceReaderBase<Integer> reader,
+            String topicName,
+            int partitionId,
+            Boundedness boundedness) {
+        PulsarPartitionSplit split = createPartitionSplit(topicName, partitionId, boundedness);
+        reader.addSplits(Collections.singletonList(split));
+        reader.notifyNoMoreSplits();
+    }
+
+    protected void pollUtilReadExpectedNumberOfRecordsAndValidate(
+            PulsarSourceReaderBase<Integer> reader,
+            TestingReaderOutput<Integer> output,
+            int expectedRecords,
+            String topicNameWithPartition)
+            throws Exception {
+        pollUntil(
+                reader,
+                output,
+                () -> output.getEmittedRecords().size() == expectedRecords,
+                "The output didn't poll enough records before timeout.");
+        reader.close();
+        verifyAllMessageAcknowledged(expectedRecords, topicNameWithPartition);
+        assertThat(output.getEmittedRecords()).hasSize(expectedRecords);
+    }
+
+    protected void pollUntil(
+            PulsarSourceReaderBase<Integer> reader,
+            ReaderOutput<Integer> output,
+            Supplier<Boolean> condition,
+            String errorMessage)
+            throws InterruptedException, TimeoutException {
+        CommonTestUtils.waitUtil(
+                () -> {
+                    try {
+                        reader.pollNext(output);
+                    } catch (Exception exception) {
+                        throw new RuntimeException(
+                                "Caught unexpected exception when polling from the reader",
+                                exception);
+                    }
+                    return condition.get();
+                },
+                Duration.ofSeconds(DEFAULT_POLLING_TIMEOUT_SECONDS),
+                errorMessage);
+    }
+
+    protected void verifyAllMessageAcknowledged(int expectedMessages, String partitionName)
```

Comment:
I wonder whether we can introduce an assertion helper class that is constructed with the `operator`. So we do not need to use inheritance to share utils